### PR TITLE
[MHLO] add EliminateRedundantConvert canonicalize pattern

### PR DESCRIPTION
**Pull request by qingyunqu**
add EliminateRedundantConvert canonicalize pattern: `convert(convert(x)) => convert(x)`

### CODE REVIEW

**Comment by joker-eph** [comment]
What about conversion like:
```
%downscale = convert %value : (f64) -> f32
%upscale = convert %downscale : (f32) -> f64
```

Or:
```
%integerRounded = convert %value : (f32) -> i32
%float = convert %integerRounded : (i32) -> f32
```

It seems to me that we can't just eliminate these as "redundant" but I don't quite get where you guard against that?

**Comment by qingyunqu** [comment]
> What about conversion like:
> 
> ```
> %downscale = convert %value : (f64) -> f32
> %upscale = convert %downscale : (f32) -> f64
> ```
> 
> Or:
> 
> ```
> %integerRounded = convert %value : (f32) -> i32
> %float = convert %integerRounded : (i32) -> f32
> ```
> 
> It seems to me that we can't just eliminate these as "redundant" but I don't quite get where you guard against that?

Yes, I think you are right. I will check the demoting convert and not to fuse them.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 1]
```diff
@@ -2140,9 +2140,55 @@ OpFoldResult ConvertOp::fold(ArrayRef<Attribute> operands) {
   return {};
 }
 
+namespace {
+
+struct EliminateRedundantConvert : public OpRewritePattern<ConvertOp> {
+  using OpRewritePattern<ConvertOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(ConvertOp op,
+                                PatternRewriter &rewriter) const override {
+    if (auto convert_op = op.operand().getDefiningOp<ConvertOp>()) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:2149]: Nit: please use early exit:
```suggestion
    auto convert_op = op.operand().getDefiningOp<ConvertOp>();
    if (!convert_op) return failure();
```

Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc]: done

**Review by joker-eph — APPROVED** [review]

**Pull request merged**